jobserver/models.py

```diff
@@ -63,14 +70,41 @@ def repo_name(self):
 
 
 class JobQuerySet(models.QuerySet):
-    def completed(self):
-        return self.filter(completed_at__isnull=False)
+    def finished(self):
+        return self.filter(
+            started=True,
+            status_code__isnull=False,
+            completed_at__isnull=False,
+        )
+
+    def failed(self):
+        ignored_states = [
+            STATE_DEPENDENCY_NOT_FINISHED,
+            STATE_DEPENDENCY_RUNNING,
+            STATE_SUCCESS,
+        ]
 
-    def in_progress(self):
-        return self.filter(started_at__isnull=False, completed_at=None)
+        return self.filter(
+            completed_at__isnull=False, status_code__isnull=False
+        ).exclude(status_code__in=ignored_states)
 
     def pending(self):
-        return self.filter(started_at=None, completed_at=None)
+        pending_states = [
+            STATE_DEPENDENCY_NOT_FINISHED,
+            STATE_DEPENDENCY_RUNNING,
+        ]
+        qwargs = Q(status_code=None) | Q(status_code__in=pending_states)
+        return self.filter(qwargs, started=False)
+
+    def running(self):
+        return self.filter(started=True, completed_at=None)
+
+    def succeeded(self):
+        return self.filter(
+            started=True,
+            completed_at__isnull=False,
+            status_code__isnull=False,
```

Comment:
This potentially needs to exclude `pending_states` too.  It's possibly that pending states never have a completed_at, but I see no downside to a belt-and-braces here,

Reply:
In `Job.is_succeeded` we're checking that the status_code is `0`, should we mirror that here instead?

Reply:
Yes... and I'd have said having two implementations of the logic for a single desired outcome is asking for trouble anyway...?

Reply:
I agree on two implementations being bad, but calling QuerySet methods in the properties is going to result in an explosion of queries.

The JobRequest list pages are paginated to 25 items.  Each of those items has 1+ related Jobs.  We render the state for each Job on the page using these properties via the `.status` property.  Worse case scenario we're doing to execute 4 queries per Job.

The current first page of Prods Jobs page has:
* 29 failed Jobs
* 19 pending Jobs
* 1 running Job

Based on the ordering of conditionals in `Job.status` that's:
* `29 * 2 = 58`
* `18 * 4 = 72`
* `1 * 3 = 3`

For a total of 133 queries 😬

Reply:
Seagull code review, please tell me to get stuffed if I'm not being helpful.

1) 133 isn't a necessarily disaster.
2) Can this be solved with a combination of `prefetch_related` and property functions that ensure that things are loaded from `_prefetched_objects_cache`?

Reply:
I might well be wrong here (please say so!), but the suggestion [here](https://github.com/opensafely/job-server/pull/180#discussion_r516842266) was to use `Job.objects.completed.filter(pk=self.pk).count() > 0`, which I don't think we could handle with `prefetch_related` but we could potentially aggregate the counts onto a QuerySet.

However, to me, that's adding indirection.  As someone querying multiple Jobs I have to remember to aggregate the correct things onto a QuerySet or get an implicit query explosion.

---

It turns out we're not actually using the QuerySet anywhere… I have a feeling this was added (and maybe used briefly?) by the JobList page filters before `JobRequest` was added and that page rebuilt 😬  On the plus side, this discussion has been great because I did sit down and interrogate all my previous decisions around this 👏